File: Editorial/22_Generate_Parentheses/Go/2_Backtracking_Keep_Candidate_Valid.go
func generateParenthesis(n int) []string {
    var answer []string
    var backtracking func(currString string, leftCount, rightCount int)
    backtracking = func(currString string, leftCount, rightCount int) {
        if len(currString) == 2*n {
            answer = append(answer, currString)
            return
        }
        if leftCount < n {
            backtracking(currString+"(", leftCount+1, rightCount)
        }
        if rightCount < leftCount {
            backtracking(currString+")", leftCount, rightCount+1)
        }
    }
    backtracking("", 0, 0)
    return answer
}

File: Editorial/22_Generate_Parentheses/TypeScript/2_Backtracking_Keep_Candidate_Valid.ts
function generateParenthesis(n: number): string[] {
    const answer: string[] = [];
    backtracking(answer, "", 0, 0, n);
    return answer;
}
function backtracking(
    answer: string[],
    curString: string,
    leftCount: number,
    rightCount: number,
    n: number,
): void {
    if (curString.length === 2 * n) {
        answer.push(curString);
        return;
    }
    if (leftCount < n) {
        curString += "(";
        backtracking(answer, curString, leftCount + 1, rightCount, n);
        curString = curString.slice(0, -1);
    }
    if (leftCount > rightCount) {
        curString += ")";
        backtracking(answer, curString, leftCount, rightCount + 1, n);
        curString = curString.slice(0, -1);
    }
}
